File: perl5/package_skill.pl
#!/usr/bin/env perl
# package_skill.py — portiert nach perl5
# Quelle: python, OpenClaw@gateway1:skills/skill-creator/scripts/package_skill.py
# auch in: OpenClaw@gateway2:skills/skill-creator/scripts/package_skill.py
# Erzeugt: 2026-08-19 durch ABSTRACTIONS_MANAGER.py

use strict;
use warnings;
use File::Find;
use File::Spec;
use File::Basename;
use Cwd 'abs_path';
use lib '.';
use quick_validate qw(validate_skill);

# Skill Packager - Creates a distributable .skill file of a skill folder
#
# Usage:
#     perl utils/package_skill.pl <path/to/skill-folder> [output-directory]
#
# Example:
#     perl utils/package_skill.pl skills/public/my-skill
#     perl utils/package_skill.pl skills/public/my-skill ./dist

sub _is_within {
    my ($path, $root) = @_;
    eval {
        my @path_parts = split '/', $path;
        my @root_parts = split '/', $root;
        for my $i (0 .. $#root_parts) {
            return 0 if $path_parts[$i] ne $root_parts[$i];
        }
        return 1;
    };
    return !$@;
}

sub package_skill {
    my ($skill_path, $output_dir) = @_;
    
    # Resolve paths
    $skill_path = abs_path($skill_path);
    
    # Validate skill folder exists
    if (!-e $skill_path) {
        print "[ERROR] Skill folder not found: $skill_path\n";
        return undef;
    }
    
    if (!-d $skill_path) {
        print "[ERROR] Path is not a directory: $skill_path\n";
        return undef;
    }
    
    # Validate SKILL.md exists
    my $skill_md = File::Spec->catfile($skill_path, "SKILL.md");
    if (!-e $skill_md) {
        print "[ERROR] SKILL.md not found in $skill_path\n";
        return undef;
    }
    
    # Run validation before packaging
    print "Validating skill...\n";
    my ($valid, $message) = validate_skill($skill_path);
    if (!$valid) {
        print "[ERROR] Validation failed: $message\n";
        print "   Please fix the validation errors before packaging.\n";
        return undef;
    }
    print "[OK] $message\n\n";
    
    # Determine output location
    my $skill_name = basename($skill_path);
    my $output_path;
    if ($output_dir) {
        $output_path = abs_path($output_dir);
        system("mkdir -p '$output_path'") == 0 or die "Failed to create output directory: $!";
    } else {
        $output_path = cwd();
    }
    
    my $skill_filename = File::Spec->catfile($output_path, "$skill_name.skill");
    
    my %EXCLUDED_DIRS = map { $_ => 1 } (".git", ".svn", ".hg", "__pycache__", "node_modules");
    
    # Create the .skill file (zip format)
    eval {
        # Collect files to add
        my @files_to_add;
        
        find(sub {
            return if $_ eq '.' || $_ eq '..';
            
            my $file_path = $File::Find::name;
            
            # Security: never follow or package symlinks.
            if (-l $file_path) {
                print "[WARN] Skipping symlink: $file_path\n";
                return;
            }
            
            my @rel_parts = split('/', File::Spec->abs2rel($file_path, $skill_path));
            my $skip = 0;
            for my $part (@rel_parts) {
                if ($EXCLUDED_DIRS{$part}) {
                    $skip = 1;
                    last;
                }
            }
            return if $skip;
            
            if (-f $file_path) {
                my $resolved_file = abs_path($file_path);
                if (!_is_within($resolved_file, $skill_path)) {
                    die "[ERROR] File escapes skill root: $file_path\n";
                }
                
                # If output lives under skill_path, avoid writing archive into itself.
                if ($resolved_file eq abs_path($skill_filename)) {
                    print "[WARN] Skipping output archive: $file_path\n";
                    return;
                }
                
                push @files_to_add, $file_path;
            }
        }, $skill_path);
        
        # Build zip command
        my $cmd = "zip -j '$skill_filename' ";
        for my $file (@files_to_add) {
            my $rel_path = File::Spec->abs2rel($file, $skill_path);
            my $arcname = File::Spec->catfile($skill_name, $rel_path);
            $cmd .= "'$file' ";
            print "  Added: $arcname\n";
        }
        
        system("$cmd >/dev/null 2>&1") == 0 or die "Failed to create zip file";
        
        print "\n[OK] Successfully packaged skill to: $skill_filename\n";
        return $skill_filename;
    };
    
    if ($@) {
        print "[ERROR] Error creating .skill file: $@\n";
        return undef;
    }
}

sub main {
    if (@ARGV < 1) {
        print "Usage: perl utils/package_skill.pl <path/to/skill-folder> [output-directory]\n";
        print "\nExample:\n";
        print "  perl utils/package_skill.pl skills/public/my-skill\n";
        print "  perl utils/package_skill.pl skills/public/my-skill ./dist\n";
        exit 1;
    }
    
    my $skill_path = $ARGV[0];
    my $output_dir = $ARGV[1] if @ARGV > 1;
    
    print "Packaging skill: $skill_path\n";
    if ($output_dir) {
        print "   Output directory: $output_dir\n";
    }
    print "\n";
    
    my $result = package_skill($skill_path, $output_dir);
    
    if ($result) {
        exit 0;
    } else {
        exit 1;
    }
}

main() if !caller;

File: javascript/package_skill.js
#!/usr/bin/env node
// package_skill.py — portiert nach javascript
// Quelle: python, OpenClaw@gateway1:skills/skill-creator/scripts/package_skill.py
// auch in: OpenClaw@gateway2:skills/skill-creator/scripts/package_skill.py
// Erzeugt: 2026-08-19 durch ABSTRACTIONS_MANAGER.py

/**
 * Skill Packager - Creates a distributable .skill file of a skill folder
 *
 * Usage:
 *     node utils/package_skill.js <path/to/skill-folder> [output-directory]
 *
 * Example:
 *     node utils/package_skill.js skills/public/my-skill
 *     node utils/package_skill.js skills/public/my-skill ./dist
 */

const fs = require('fs');
const path = require('path');
const { program } = require('commander');
const AdmZip = require('adm-zip');
const { validateSkill } = require('./quick_validate');

function isWithin(filePath, rootPath) {
    const relative = path.relative(rootPath, filePath);
    return !relative.startsWith('..') && !path.isAbsolute(relative);
}

function packageSkill(skillPath, outputDir = null) {
    // Resolve paths
    skillPath = path.resolve(skillPath);

    // Validate skill folder exists
    if (!fs.existsSync(skillPath)) {
        console.error(`[ERROR] Skill folder not found: ${skillPath}`);
        return null;
    }

    if (!fs.statSync(skillPath).isDirectory()) {
        console.error(`[ERROR] Path is not a directory: ${skillPath}`);
        return null;
    }

    // Validate SKILL.md exists
    const skillMd = path.join(skillPath, 'SKILL.md');
    if (!fs.existsSync(skillMd)) {
        console.error(`[ERROR] SKILL.md not found in ${skillPath}`);
        return null;
    }

    // Run validation before packaging
    console.log('Validating skill...');
    const validationResult = validateSkill(skillPath);
    if (!validationResult.valid) {
        console.error(`[ERROR] Validation failed: ${validationResult.message}`);
        console.error('   Please fix the validation errors before packaging.');
        return null;
    }
    console.log(`[OK] ${validationResult.message}\n`);

    // Determine output location
    const skillName = path.basename(skillPath);
    let outputPath;
    if (outputDir) {
        outputPath = path.resolve(outputDir);
        fs.mkdirSync(outputPath, { recursive: true });
    } else {
        outputPath = process.cwd();
    }

    const skillFilename = path.join(outputPath, `${skillName}.skill`);

    const EXCLUDED_DIRS = new Set(['.git', '.svn', '.hg', '__pycache__', 'node_modules']);

    // Create the .skill file (zip format)
    try {
        const zip = new AdmZip();

        // Walk through the skill directory
        function walkDirectory(currentPath) {
            const items = fs.readdirSync(currentPath, { withFileTypes: true });
            
            for (const item of items) {
                const itemPath = path.join(currentPath, item.name);
                
                // Security: never follow or package symlinks.
                if (fs.lstatSync(itemPath).isSymbolicLink()) {
                    console.warn(`[WARN] Skipping symlink: ${itemPath}`);
                    continue;
                }

                const relativeParts = path.relative(skillPath, itemPath).split(path.sep);
                if (relativeParts.some(part => EXCLUDED_DIRS.has(part))) {
                    continue;
                }

                if (item.isDirectory()) {
                    walkDirectory(itemPath);
                } else if (item.isFile()) {
                    const resolvedFile = path.resolve(itemPath);
                    if (!isWithin(resolvedFile, skillPath)) {
                        console.error(`[ERROR] File escapes skill root: ${itemPath}`);
                        return null;
                    }
                    
                    // If output lives under skillPath, avoid writing archive into itself.
                    if (path.resolve(resolvedFile) === path.resolve(skillFilename)) {
                        console.warn(`[WARN] Skipping output archive: ${itemPath}`);
                        continue;
                    }

                    // Calculate the relative path within the zip.
                    const arcname = path.join(skillName, path.relative(skillPath, itemPath));
                    zip.addLocalFile(itemPath, path.dirname(arcname));
                    console.log(`  Added: ${arcname}`);
                }
            }
        }

        walkDirectory(skillPath);
        zip.writeZip(skillFilename);

        console.log(`\n[OK] Successfully packaged skill to: ${skillFilename}`);
        return skillFilename;

    } catch (e) {
        console.error(`[ERROR] Error creating .skill file: ${e.message}`);
        return null;
    }
}

function main() {
    program
        .arguments('<skill-path> [output-directory]')
        .action((skillPath, outputDir) => {
            console.log(`Packaging skill: ${skillPath}`);
            if (outputDir) {
                console.log(`   Output directory: ${outputDir}`);
            }
            console.log();

            const result = packageSkill(skillPath, outputDir);

            if (result) {
                process.exit(0);
            } else {
                process.exit(1);
            }
        });

    if (process.argv.length < 3) {
        console.log('Usage: node utils/package_skill.js <path/to/skill-folder> [output-directory]');
        console.log('\nExample:');
        console.log('  node utils/package_skill.js skills/public/my-skill');
        console.log('  node utils/package_skill.js skills/public/my-skill ./dist');
        process.exit(1);
    }

    program.parse(process.argv);
}

if (require.main === module) {
    main();
}

module.exports = { packageSkill };
